### PR TITLE
feat: k8s patch scaling component

### LINKED ISSUE #162: Kubernetes Patch Scaling Component
kubernetes patch - update manifest

**ScalingComponent** (kubernetes-patch)
```yaml
kind: ScalingComponent
id: k8s_istio_vs
component_kind: kubernetes-patch
metadata:
  api_server_endpoint: "https://localhost"
```

**ScalingPlan**
```yaml
kind: ScalingPlan
id: scaling_plan_k8s_istio_vs_autoscaling
metadata:
  title: "Scaling Plan for K8S istio VirtualService Autoscaling"
  cool_down: 300 # seconds
  interval: 10000 # milliseconds
plans:
  - id: scale-out-plan-1
    description: "Scale out"
    expression: >
      get({
        metric_id: 'metric_id_1',
        name: 'name_1',
        stats: 'max',
        period_sec: 10
      }) >= 100
    priority: 1
    scaling_components:
      - component_id: k8s_istio_vs
        namespace: kube-namespace
        name: istio_vs
        manifest:
          apiVersion: networking.istio.io/v1beta1
          kind: VirtualService
          spec:
            hosts:
              - "*"
            gateways:
              - istio-gateway
            http:
              - match:
                - uri:
                     prefix: /
                   port: 9090
                 route:
                   - destination:
                        host: prometheus-community-kube-prometheus.istio-system.svc.cluster.local
                        port:
                          number: 9090
```

### PR DESCRIPTION
**Pull request by Ari-suhyeon**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] DevOps process (non-breaking change which improves efficiency and reliability for CI/CD)
- [ ] Documentation only

## What this PR does / why we need it:

kubernetes patch - update manifest

**ScalingComponent** (kubernetes-patch)
```yaml
kind: ScalingComponent
id: k8s_istio_vs
component_kind: kubernetes-patch
metadata:
```

**ScalingPlan**
```yaml
kind: ScalingPlan
id: scaling_plan_k8s_istio_vs_autoscaling
metadata:
  title: "Scaling Plan for K8S istio VirtualService Autoscaling"
  cool_down: 30 # seconds
  interval: 10000 # milliseconds
plans:
  - id: scale-out-plan-1
    description: "Scale out"
    expression: >
      get({
        metric_id: 'metric_id_1',
        name: 'name_1',
        stats: 'max',
        period_sec: 10
      }) >= 100
    priority: 1
    scaling_components:
      - component_id: k8s_istio_vs
        namespace: kube-namespace
        name: istio_vs
        manifest:
          apiVersion: networking.istio.io/v1beta1
          kind: VirtualService
          spec:
            hosts:
              - "*"
            gateways:
              - istio-gateway
            http:
              - match:
                - uri:
                     prefix: /
                   port: 9090
                 route:
                   - destination:
                        host: prometheus-community-kube-prometheus.istio-system.svc.cluster.local
                        port:
                          number: 9090
```

## Which issue/s this PR fixes

fixes #162 

## How Has This Been Tested?

* moon run tested - GKE
* Unit Test has been tested.

### CODE REVIEW

File: core/wave-autoscale/src/scaling_component/k8s_patch.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                .get("api_server_endpoint")
+                .map(|api_server_endpoint| api_server_endpoint.to_string());
+        let ca_cert = metadata
+                .get("ca_cert")
+                .map(|ca_cert| ca_cert.to_string());
+        let Ok(client) = self.get_client(api_server_endpoint, ca_cert, Some(namespace.to_string())).await else {
+            return Err(anyhow::anyhow!("cannot create kubernetes client"));
+        };
+
+        println!("manifest:\n {:#?}", manifest);
```
Comment (pueding) [inline core/wave-autoscale/src/scaling_component/k8s_patch.rs:69]: Remove the println!

File: core/wave-autoscale/src/scaling_component/k8s_patch.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        let api_version = manifest.get("apiVersion").unwrap();
+        let api_group = api_version.as_str().unwrap().split("/").nth(0).unwrap();
+        let kind = manifest.get("kind").unwrap().as_str().unwrap();
+
+        let apigroup = discovery::group(&client, api_group).await.unwrap();
+        let (ar, caps) = apigroup.recommended_kind(kind).unwrap();
```
Comment (pueding) [inline core/wave-autoscale/src/scaling_component/k8s_patch.rs:75]: handle the error before the unwraps

File: core/wave-autoscale/src/scaling_component/k8s_patch.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        let api: Api<DynamicObject> = Api::namespaced_with(client, namespace, &ar);
+
+        let patch_params = PatchParams::apply("wave-autoscale");
+        let patch_params = PatchParams::force(patch_params);
```
Comment (pueding) [inline core/wave-autoscale/src/scaling_component/k8s_patch.rs:79]: Add a comment why use apply and force together

File: core/wave-autoscale/src/scaling_component/k8s_patch.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+        let result = api
+            .patch(name, &patch_params, &Patch::Apply(manifest))
+            .await;
+
+        println!("{:#?}", result);
```
Comment (pueding) [inline core/wave-autoscale/src/scaling_component/k8s_patch.rs:85]: Remove

**Review by pueding — CHANGES_REQUESTED** [review]

File: core/wave-autoscale/src/scaling_component/k8s_patch.rs (hunk 3 above)
Comment (Ari-suhyeon) [reply core/wave-autoscale/src/scaling_component/k8s_patch.rs:79]: There was a conflict with terraform, so added the force option.
`"Apply failed with 1 conflict: conflict with \"HashiCorp\" using apps/v1: .spec.replicas"`

**Review by pueding — APPROVED** [review]

**Pull request merged**